nameserver/server.go

```diff
@@ -32,7 +20,94 @@ func makeDNSFailResponse(r *dns.Msg) *dns.Msg {
 	return m
 }
 
-func queryHandler(lookups []Lookup) dns.HandlerFunc {
+type DNSServer struct {
+	config *dns.ClientConfig
+	zone   Zone
+	iface  *net.Interface
+	port   int
+	udpSrv *dns.Server
+	tcpSrv *dns.Server
+}
+
+// Creates a new DNS server with a given config
+func NewDNSServerWithConfig(config *dns.ClientConfig, zone Zone, iface *net.Interface, port int) (*DNSServer, error) {
+	return &DNSServer{
+		config: config,
+		zone:   zone,
+		iface:  iface,
+		port:   port,
+	}, nil
+}
+
+// Creates a new DNS server
+func NewDNSServer(zone Zone, iface *net.Interface, port int) (*DNSServer, error) {
+	config, err := dns.ClientConfigFromFile("/etc/resolv.conf")
+	if err != nil {
+		return nil, err
+	}
+	return NewDNSServerWithConfig(config, zone, iface, port)
+}
+
+// Start the DNS server
+func (s *DNSServer) Start() error {
+	mdnsClient, err := NewMDNSClient()
+	checkFatal(err)
+
+	ifaceName := "default interface"
+	if s.iface != nil {
+		ifaceName = s.iface.Name
+	}
+	Info.Printf("Using mDNS on %s", ifaceName)
+	err = mdnsClient.Start(s.iface)
+	checkFatal(err)
+
+	// create two DNS request multiplexerers, depending on the protocol used by clients
+	// (we use the same protocol for asking upstream servers)
+	mux := func(client *dns.Client) *dns.ServeMux {
+		m := dns.NewServeMux()
+		m.HandleFunc(LOCAL_DOMAIN, s.queryHandler([]Lookup{s.zone, mdnsClient}))
+		m.HandleFunc(RDNS_DOMAIN, s.rdnsHandler([]Lookup{s.zone, mdnsClient}, client))
+		m.HandleFunc(".", s.notUsHandler(client))
+		return m
+	}
+
+	mdnsServer, err := NewMDNSServer(s.zone)
+	checkFatal(err)
+
+	err = mdnsServer.Start(s.iface)
+	checkFatal(err)
+
+	address := fmt.Sprintf(":%d", s.port)
+	s.udpSrv = &dns.Server{Addr: address, Net: "udp", Handler: mux(&dns.Client{Net: "udp", UDPSize: UDPBufSize})}
+	s.tcpSrv = &dns.Server{Addr: address, Net: "tcp", Handler: mux(&dns.Client{Net: "tcp"})}
```

Comment:
All of this initialization stuff (ie, `mux()`, `s.udpSrv`, `s.tcpSrv`) has been moved to `NewDNSServer()` in zettio/weave#394. The `Start()` method should just do that: start the server.